Add connection tables to details panel

tomwilkie opened the pull request:
Part of #680, Fixes #378, Fixes #1111

This change also:
- Makes RenderableNode.Children a set of RenderableNodes, not Nodes
- <s>Puts pseudo nodes back into the container's view (makes more sense to show connections now we have shapes)</s> (This needed more work - #1123)

TODO:
- [x] make it sort by count, by default
- [x] change table ID as per davkal's request
- [x] make connection work to/from the internet node
- [x] figure out why the memoisation causes the test to fail
- [x] testing / coverage
- [x] add real nodes to the table
- [x] figure out why inbound count != outbound count for self-connections
- [x] better table for the internet node
- <s>use reverse dns info</s> (#1121)

paulbellamy commented on report/id_list.go:
```diff
@@ -35,6 +33,10 @@ func (a IDList) Merge(b IDList) IDList {
+
+// Intersection returns the intersection of a and b
+func (a IDList) Intersection(b IDList) IDList {
+	return IDList(StringSet(a).Intersection(StringSet(b)))
```
nice

paulbellamy commented on render/detailed/node.go:
```diff
@@ -157,5 +167,140 @@ func children(n render.RenderableNode) []NodeSummaryGroup {
+				return
+			}
+
+			intersection := child.Adjacency.Intersection(myEndpoints)
```
nice

paulbellamy commented on render/detailed/node.go:
```diff
@@ -157,5 +167,140 @@ func children(n render.RenderableNode) []NodeSummaryGroup {
+				remotes[key] = remotes[key] + 1
+			}
+		}
+	}
```
Seems very similar to the above. possible to extract/combine/refactor?

davkal commented:
How about making connections a toplevel node field:
```
{
metadata,
metrics,
connections: [
  {
    id: "out",
    columns: [{id: "id", label: "Remote"}, {id: "port", label: "Port"}, {id: "count", label: "Label"}}],
    items:[{id: <nodeId>, port: <port>, count: <count>}, ...]
  },
  {id: in, ...}
]}
```

tomwilkie commented:
Sorry should have WIP marker.  Yes I hope to combined lots of this code, once I get it working.

tomwilkie commented:
@davkal I eventually want the rows to link to other objects in the topology (ie link to the other container talking to this one) like the other children tables.  Links will always be within the current view.  Can you update your datastructure for this?  I'm hoping links will always be possible, but you might want to make them optional.

paulbellamy commented:
just reading a bit for curiosity.

davkal commented:
Alternatively we can model connections as a nodes table, i.e. wrapping ports as metadata (typed as int to make it sortable as a number) and counts wrapped as a metric:
```
{
metadata,
metrics,
children,
connections: [{
    id: "out",
    label: "Outgoing Connections",
    columns: [{id: "label", label: "Remote"}, {id: "port", label: "Port"}, {id: "count", label: "Count"}}],
    nodes: [{
      id: <nodeId>,
      label: <nodeLabel>,
      topologyId: <topologyId>,
      metadata: [{port: <port>, type: "number"}]
      metrics: [{count: <count>}]
    }, ...]
  },
  {id: in, ...}

davkal commented:
`topologyId: "endpoint-incoming"` needs to be `id: "endpoint-incoming"`. `topologyId` is something like `processes` that will be used together with a node's `id` to build a deeplink.

tomwilkie commented on app/api_topology_test.go:
```diff
@@ -54,27 +54,6 @@ func TestAll(t *testing.T) {
 }
 
-func TestAPITopologyContainers(t *testing.T) {
-	ts := topologyServer()
```
I dropped this test as rendered containers contain a bunch of extra special casing (image names in rank etc) and its verbose testing that doesn't add anything.

paulbellamy commented on app/api_topology_test.go:
```diff
@@ -112,6 +91,7 @@ func TestAPITopologyHosts(t *testing.T) {
 	defer ts.Close()
 	is404(t, ts, "/api/topology/hosts/foobar")
 	{
+		render.ResetCache() // TODO figure out why this is needed.
```
Why *is* this needed? each report should have a different ID, yeah? ID collisions in the renderers, or reports?

tomwilkie replied:
I was getting errors without it.  I don't know the root of those errors.

tomwilkie replied:
Pretty sure its solved by #1088, will rebase when thats merged in.

paulbellamy commented on render/detailed/node.go:
```diff
@@ -157,5 +165,144 @@ func children(n render.RenderableNode) []NodeSummaryGroup {
+	remotePortKey = "Remote Port"
+	localPortKey  = "Local Port"
+	countKey      = "Count"
+	numberWang    = "number"
```
wat

tomwilkie replied:
https://www.youtube.com/watch?v=qjOZtWZ56lc

paulbellamy commented on render/detailed/node.go:
```diff
@@ -157,5 +165,144 @@ func children(n render.RenderableNode) []NodeSummaryGroup {
+			intersection := child.Adjacency.Intersection(myEndpoints)
+			if len(intersection) <= 0 {
+				return
+			}
```
don't really need this, the below loop will just skip.

paulbellamy commented:
still quite a bit of tricky string-munging code shared between the incoming connections and outgoing connection aggregators.

paulbellamy commented:
`Connections Fr...` is not a great table name.
<img width="435" alt="screen shot 2016-02-25 at 16 17 40" src="https://cloud.githubusercontent.com/assets/250199/13325823/5ef13ba8-dbdb-11e5-888a-5e0632399172.png">

paulbellamy commented:
I guess no luck on converting the endpoints to their respective nodes in the relevant topology?

tomwilkie commented:
> I guess no luck on converting the endpoints to their respective nodes in the relevant topology?

Time constraint I'm afraid.  Will do it next sprint.

paulbellamy commented:
Aside from a few naming things and refactoring (if possible), this LGTM.

tomwilkie commented:
@paulbellamy I think I've addressed all the feedback, and got it working for the internet node.  Its a bit of a hack.  I think we should punt this form 0.13 and finish it properly.  Thoughts?

tomwilkie commented:
Also think the internet is going to need a custom table (for outgoing connections):

<img width="449" alt="screen shot 2016-02-25 at 19 10 27" src="https://cloud.githubusercontent.com/assets/444037/13331035/827e0e4e-dbf3-11e5-9922-96809024cf5f.png">

tomwilkie commented:
Going to hold this until #566 is merged.  Also need @pidster input on the table for the internet node(s).

paulbellamy commented on render/id.go:
```diff
@@ -14,6 +24,11 @@ func MakeEndpointID(hostID, addr, port string) string {
+// MakePseudoEndpointID makes a pseudo endpoint node ID for rendered nodes.
+func MakePseudoEndpointID(hostID, addr, port string) string {
+	return makeID("pseudo-endpoint", hostID, addr, port)
+}
```
Seems an odd differentiation between `pseudo` nodes and `pseudo-endpoint` nodes.

tomwilkie replied:
indeed, I just couldn't think of a better name.  Any suggestions?

paulbellamy replied:
Not the naming. It seems like an odd way to slice the distinction.

tomwilkie replied:
Well the whole point is to have the pseudo nodes have endpoints as their children.  I guess there are two alternatives:
- introduce the pseudo nodes in the process mapping instead of the endpoint mapper
- make endpoint mapper add the 'report' endpoint nodes (from the selector) as children of the endpoints & pseudo nodes, and make the connection table code look for them instead

Can you think of any other alternatives?  Which do you prefer?

paulbellamy replied:
> introduce the pseudo nodes in the process mapping instead of the endpoint mapper

That would be more consistent with how the "uncontained" pseudo nodes are handled, no?

paulbellamy commented on render/mapping.go:
```diff
@@ -50,15 +51,31 @@ func theInternetNode(m RenderableNode) RenderableNode {
 	return node
 }
 
+// RemapEndpointIDs remaps endpoints to have the right id format.
```
This comment is unhelpful. What format does it expect them to have? what is the `right` format?

paulbellamy commented on render/mapping.go:
```diff
@@ -118,7 +137,9 @@ func MapEndpointIdentity(m RenderableNode, local report.Networks) RenderableNode
 
-	return RenderableNodes{id: NewRenderableNodeWith(id, major, minor, rank, m)}
+	node := NewRenderableNodeWith(id, major, minor, "", m)
+	node.Children = node.Children.Add(m)
```
eh? so endpoints are children of themselves? That seems odd?

tomwilkie replied:
This is the endpoint being added as a child of the pseudo endpoint.

paulbellamy replied:
ah

paulbellamy commented on render/mapping.go:
```diff
@@ -155,10 +176,9 @@ func MapContainerIdentity(m RenderableNode, _ report.Networks) RenderableNodes {
-		rank, _  = m.Latest.Lookup(docker.ImageID)
 	)
 
-	node := NewRenderableNodeWith(id, major, minor, rank, m)
```
why is rank going away?

tomwilkie replied:
It gets over written by containerWithImageNameRenderer anyway.

paulbellamy commented on render/mapping.go:
```diff
@@ -453,14 +473,9 @@ func MapEndpoint2Process(n RenderableNode, _ report.Networks) RenderableNodes {
-	}
-
-	// Don't propogate non-internet pseudo nodes
+	// Propogate pseudo node
```
Propagate

tomwilkie replied:
Done.

paulbellamy commented on render/detailed/connections.go:
```diff
@@ -0,0 +1,197 @@
+package detailed
```
This file may be easier to read if you flip the order so the primary functions (`makeIncomingConnectionsTable` and `makeOutgoingConnectionsTable`) were first, then the helpers afterwards.

paulbellamy replied:
Also, it seems like the functions have a bit of redundancy in the names `buildThis` and `makeThat`

tomwilkie replied:
Done & done.

paulbellamy commented:
General approach seems fine. A couple naming and tidiness things.

rade commented on render/detailed/connections.go:
```diff
@@ -0,0 +1,197 @@
+					ID:       "foo",
+					Value:    row.localAddr,
+					Datatype: number,
+				})
```
@tomwilike as the author, and @davkal @paulbellamy as the reviewers: Can you shed any light on

1. the significance of "foo"
2. why the Datatype is 'number' when the Value is an IP address (and in more recent versions an IP address or DNS name with IP address in parentheses)? Won't this cause `parseFloat` in https://github.com/weaveworks/scope/pull/1017/files#diff-37e17af975f5417e26f6036ba3624af9R55 to return "NaN"?

paulbellamy replied:
> The significance of MetadataRow.ID == "foo"

Just has to match the Column.ID == "foo" above.

> Won't this cause parseFloat to return "NaN".

No.

rade replied:
> Just has to match the Column.ID == "foo" above.

Well yes, I figured that. But why is it called "foo" when all the other IDs are named sensibly?

> No.

Why not? And why do we want to set the type to 'number' here when the value *isn't* a number?

paulbellamy replied:
> And why do we want to set the type to 'number' here when the value isn't a number?

Because we want them sorted numerically, not lexicographically.

rade replied:
> Because that is not what parseFloat does. Try it: `parseFloat("127.0.0.1")`

I did try it before my original comment. Try: `parseFloat("cloud.weave.works (52.3.116.14)")`.

rade replied:
Sorting IP addresses by converting them to a float?!?! You are having a laugh.
```
> parseFloat("10.0.121.1")
10
> parseFloat("10.0.118.1")
10
```

paulbellamy replied:
> Sorting IP addresses by converting them to a float?!?!

Pull requests are welcome. Also, a better example of the behaviour is:
```
> parseFloat("192.168.12.13")
192.168
```
It is probably (almost definitely) older than the reverse dns lookup, so wouldn't have ever handled `cloud.weave.works (52.3.116.14)`.

rade replied:
> it is probably (almost definitely) older than the reverse dns lookup

Right, I said that much: "in more recent versions"

Regarding the 'foo'... given that the other `MetadataRow.ID` values in this code are 'port' and 'count', which match the column labels, why not 'remote'?

paulbellamy replied:
Sure, that would make sense.

rade replied:
I have filed #1869 for the sorting problem.

rade replied:
Looks like @foot is doing the de-foo'ing in  #1874.

foot replied:
Yeah I had to reference the col id in the js so changed the name.

I dropped `dataType: number` too as I didn't consider the creative applications of parseFloat, I'd start introducing new dataTypes though, `number|ip|ips`
